File: Cores/OSOL.Extremum.Core.DotNet/Random/GoRN.cs
using System;
using OSOL.Extremum.Core.DotNet.Random.Distributions;

namespace OSOL.Extremum.Core.DotNet.Random
{
    public static class GoRN
    {
        private class Core: DiscreteUniform, ContinuousUniform, Normal
        {
            private System.Random _seed;

            public Core(int seed)
            {
                this._seed = new System.Random(seed);
            }

            public Core()
            {
                this._seed = new System.Random();
            }

            public int getDiscreteUniform(int min, int max) => _seed.Next(min, max + 1);

            public double getContinuousUniform(double min, double max) => min + _seed.NextDouble() * (max - min);

            public double getNormal(double mu, double sigma)
            {
                double x = getContinuousUniform(-1.0, 1.0);
                double y = getContinuousUniform(-1.0, 1.0);
                double s = x * x + y * y;

                while (s > 1)
                {
                    x = getContinuousUniform(-1.0, 1.0);
                    y = getContinuousUniform(-1.0, 1.0);
                    s = x * x + y * y;
                }

                double z = x * Math.Sqrt(-2 * Math.Log(s) / s);
                return mu + sigma * z;
            }
        }
        
        private static Core _core = new Core();
    }
}

File: Cores/JVM/src/main/scala/OSOL/Extremum/Cores/JVM/Random/GoRN.scala
package OSOL.Extremum.Cores.JVM.Random

import OSOL.Extremum.Cores.JVM.Random.Distributions._
import OSOL.Extremum.Cores.JVM.Random.Distributions.{ContinuousUniform, DiscreteUniform, Normal}

object GoRN
  extends DiscreteUniform
    with ContinuousUniform
    with Normal {

  private val core = new scala.util.Random()

  def resetCore(seed: java.lang.Long): Unit = core.setSeed(seed)

  override def getDiscreteUniform(min: java.lang.Integer, max: java.lang.Integer): java.lang.Integer = min + core.nextInt(max - min + 1)

  override def getContinuousUniform(min: java.lang.Double, max: java.lang.Double): java.lang.Double = min + core.nextDouble() * (max - min)

  override def getNormal(mu: java.lang.Double, sigma: java.lang.Double): java.lang.Double = {

    var x = getContinuousUniform(-1.0, 1.0)
    var y = getContinuousUniform(-1.0, 1.0)
    var s = x * x + y * y

    while (s > 1) {
      x = getContinuousUniform(-1.0, 1.0)
      y = getContinuousUniform(-1.0, 1.0)
      s = x * x + y * y
    }

    val z = x * math.sqrt(-2 * math.log(s) / s)
    mu + sigma * z

  }

  def getFromSeries[T](data: Seq[T], n: java.lang.Integer, withReturn: java.lang.Boolean): Seq[T] =
    withReturn match {
      case java.lang.Boolean.TRUE => Seq.fill(n)(getDiscreteUniform(0, data.size - 1)).map(x => data(x))
      case java.lang.Boolean.FALSE => data.map(x => (x, getContinuousUniform(0.0, 1.0))).sortBy(_._2).map(_._1).take(n)
    }

}